README
NAME
    VCS - Library for generic Version Control System access in Perl

SYNOPSIS
        use VCS;
        $file = VCS::File->new($ARGV[0]);
        print $file->url, ":\n";
        for $version ($file->versions) {
            print $version->version,
                  ' was checked in by ',
                  $version->author,
                  "\n";
        }

DESCRIPTION
    `VCS' is an API for abstracting access to all version control systems
    from Perl code. This is achieved in a similar fashion to the `DBI' suite
    of modules. There are "container" classes, `VCS::Dir', `VCS::File', and
    `VCS::Version', and "implementation" classes, such as `VCS::Cvs::Dir',
    `VCS::Cvs::File', and `VCS::Cvs::Version', which are subclasses of their
    respective "container" classes.

    The container classes are instantiated with URLs. There is a URL scheme
    for entities under version control. The format is as follows:

        vcs://localhost/VCS::Cvs/fs/path/?query=1

    The "query" part is ignored for now. The path must be an absolute path,
    meaningful to the given class. The class is an implementation class,
    such as `VCS::Cvs'.

    The "container" classes work as follows: when the `new' method of a
    container class is called, it will parse the given URL, using the
    `VCS->parse_url' method. It will then call the `new' of the
    implementation's appropriate container subclass, and return the result.
    For example,

        VCS::Version->new('vcs://localhost/VCS::Cvs/fs/path/file/1.2');

    will return a `VCS::Cvs::Version'.

    An implementation class is recognised as follows: its name starts with
    `VCS::', and `require "VCS/Classname.pm"' will load the appropriate
    implementation classes corresponding to the container classes.

VCS METHODS
  VCS->parse_url

    This returns a four-element list:

        ($hostname, $classname, $path, $query)

    For example,

        VCS->parse_url('vcs://localhost/VCS::Cvs/fs/path/file/1.2');

    will return

        (
            'localhost',
            'VCS::Cvs',
            '/fs/path/file/1.2',
            ''
        )

    This is mostly intended for use by the container classes, and its
    interface is subject to change.

  VCS->class_load

    This loads its given implementation class.

    This is mostly intended for use by the container classes, and its
    interface is subject to change.

VCS::* METHODS
    Please refer to the documentation for the VCS::Dir manpage, the
    VCS::File manpage, and the VCS::Version manpage; as well as the
    implementation specific documentation as in the VCS::Cvs manpage, the
    VCS::Rcs manpage.

AVAILABILITY
    VCS.pm and its friends are available from CPAN. There is a web page at:

        http://www.astray.com/VCS/

    as well as a sourceforge project page at:

        http://sourceforge.net/projects/vcs/

MAILING LIST
    There is currently a mailing list about VCS. Go to the following webpage
    to subscribe to it:

        http://www.astray.com/mailman/listinfo/vcs

    There is a web archive of the mailing list at:

        http://www.astray.com/pipermail/vcs/

    General queries should be made directly to the mailing list.

COPYRIGHT
    Copyright (c) 1998-2001 Leon Brocard. All rights reserved. This program
    is free software; you can redistribute it and/or modify it under the
    same terms as Perl itself.

SEE ALSO
    the VCS::Cvs manpage, the VCS::Dir manpage, the VCS::File manpage, the
    VCS::Rcs manpage, the VCS::Version manpage.



Changes
Revision history for Perl extension VCS.

0.16  Sun Aug 22 10:04:00 PDT 2010
	- Fix META.yml

0.15  Sun Aug 22 01:37:00 PDT 2010
	- VCS is obsolete, use VCI instead.

0.14  Tue Sep 14 11:21:16 BST 2004
	- Fixed a test, got shocked at how long this has been dormant.
	  (thanks to richard for reporting the test failure on some
	   new version of CVS)

0.13  Fri Jan  2 10:18:44 GMT 2004
 	- Renamed the test files from ,v to ,v_for_testing as reports
	  indicated this was causing problems.
	- Added patch from Richard Clamp to handle directories with
	  non-versioned files more gracefully.
	- Moved CVS fully onto faith.mccarroll.org.uk away from
	  sourceforge. Please contact me if you need access.
	
0.12  Sun Dec 21 20:39:45 GMT 2003
	- Added POD for the changes to tags(), sorry folks forgot
	  to do it earlier.  
	- p.s. btw, this is now no longer hosted on SF's CVS, if you
	  want CVS access let me know as i'm upping the pace dramatically
	  on this module so it should at long last start to go somewhere.
	
0.11  Sun Dec 21 17:17:45 GMT 2003
	- Applied patch from Pierre Denis (Thanks Leon!), which sorted
          out a problem with testing Rcs from a CVS checkout.
	- Started to split tags() into tags_hash and tags_array, the
	  later maintaining order.

0.10  Wed Feb  5 16:10:25 GMT 2003
	- Applied patch from Leon Broacard (Thanks Leon!), which sorted
	  out a problem with pservers and absolute paths.
	- Updated my email address (word of wisdom, never change your
	  email address its too painful)

0.09  Sun Jan 19 13:12:39 GMT 2003
	- Applied patch from Slaven Rezic (Thanks Slaven!), which ...
		- cleaned up the _CMD variables making them global
		- %LOG_CACHE also global (this allows a quick hack
		  to invalidate the CACHE)
		- use strict applied to VCS::Rcs::Version
		- text() now uses PATH instead of NAME
	- Broke out the tests into seperate files for General (00VCS.t),
	  RCS (01Rcs.t) and CVS (01Cvs.t)
	- Removed dependency on the UNIX command line tools for making
	  a test directory. 
	- Added skips for 01Rcs.t and 01Cvs.t if rcs or cvs are not
	  available.
	- Made the testing system a little more Win32 friendly.	 

0.08  Thu Jan 16 16:19:37 GMT 2003
	- Added a tags() method to VCS::Dir (see POD for more details), 
	  there are currently only tests for this function in the CVS
	  testsuite, this will change in the next version which should be
	  an overhaul of the entire test suite. There should also be a 
	  nice example of how powerful tags() on VCS::Dir can be in the
	  next version.

0.07  Tue Jan 14 UNKNOWN GMT 2003
	- Removed dependencies on uudecode, gzip and tar in the tests in
	  response to failed test report from CPAN testers (Thanks
	  CPAN Testers!) This has been done by moving the test
	  directory structure into t/.
	- Moved test.pl to t/VCS.t

0.06  Mon Jan 13 UNKNOWN GMT 2003
	- Changed test.pl to use Test::More
	- Added tags() method to VCS::File implementations in CVS, RCS, HMS.

0.05  Sun Jul 28 01:09:44 BST 2002
	- VCS URLs, eliminating problems with which kind of VCS to use.
	- Exceptions on failed calls to "new" instead of returning undef. This
	  enables calling code to assume it "just worked", and also when it
	  fails, information is available (from the message) as to just what
	  went wrong.
	- An "init" method, callable by "new" in subclasses, cutting down on
	  redundant code in subclasses.
	- A "path" method to use instead of "name", which was potentially
	  ambiguous. The data for this is setup in "init", so subclasses
	  should not need to touch.
	- Because of the above, subclasses like e.g. VCS::X::Dir must inherit
	  from VCS::Dir. If they wish to inherit from VCS::X, they should set
	  @ISA to qw(VCS::X VCS::Dir).

0.04  Oct 28 2001
        - [now in SourceForge CVS]
        - Added VCS::Hms
	- Added multitude of changes from Peter Prymmer:
        - modified the MANIFEST to include the various Hms*
          files as well as the new VCS_dev.pod doc and a Imp_pod.tpl
          file
	- added pod to VCS/Rcs.pm and similar pod to VCS/Hms.pm.
          He turned the generalization of pod into the Imp_pod.tpl pod
          template for other VCS::* developers to borrow from.
        - touched up other pod to render it `podchecker -warnings
          -warnings` using podchecker from a test installation of perl
          5.7.2 (perl@12494 kit from jhi). This mostly entailed
          s/\$object->method/\$object-E<gt>method/ in =head[12]
          headings. He added gratuitous =cut's to most pods (even
          though podchecker does not seem to mind their absence).
        - He turned one of the example programs into a MakeMaker
          extractable *.PL script.  Specifically he turned
          examples/diff-hist into examples/diff-hist.PL and made
          appropriate changes to the Makefile.PL. He has only touched
          the one program and wanted to solicit comment on this
          installation technique: note that by running `make install`
          the blib/scripts/diff-hist script will be installed into
          your @INC in such a way that `diff-hist` will lie in you
          $PATH.  Is this something worth doing with the other
          programs in the examples/* directory? Should the examples
          be given other more distinctive names such as vcsdiff-hist,
          i.e. with a leading 'vcs' name?

0.03  Thu May 27 14:31:20 BST 1999
        - more documentation
	- a couple of bug fixes

0.02  Mon Feb 22 17:11:06 1999
        - lots of bugfixes and documentation
	  updates and clarifications
        - fixed bug in the caching which meant
	  that we got almost no hits
	- fixed bug which caused us to do nasty
	  things if VCS::Rcs saw any locked files

0.01  Mon Nov 16 12:41:10 1998
	- first released version, with powerful
          vcsweb.cgi tool




VCS/Cvs.pm
package VCS::Cvs;

use strict;
use vars qw($VERSION);
use Cwd qw(cwd);
use File::Spec::Functions qw(splitpath);
use VCS::Cvs::Dir;
use VCS::Cvs::File;
use VCS::Cvs::Version;

$VERSION = '0.06';

my $LOG_CMD = "cvs log";

my %LOG_CACHE;

sub _boiler_plate_info {
    my ($self, $what) = @_;
    my ($header, $log) = $self->_split_log($self->{VERSION});
    my $rev_info = $self->_parse_log_rev($log);
    $rev_info->{$what};
}

sub _split_log {
    my ($self, $version) = @_;
    my $log_text;
    my $cache_id = $self->url;
    unless (defined($log_text = $LOG_CACHE{$cache_id})) {

        my $cwd = cwd();
        my($volume, $path, $file) = splitpath($self->path);
        chdir $path;

        my $cmd =
            $LOG_CMD .
            " " . $file . " 2>/dev/null |";
#warn "cmd: $cmd\n";
        $LOG_CACHE{$cache_id} = $log_text = $self->_read_pipe($cmd);

        chdir $cwd;
    }
    my @sections = split /\n[=\-]+\n/, $log_text;
    @sections = ($sections[0], grep { /^revision $version(?:\s+locked by.*?)\n/ } @sections)
        if $version;
#map { print "SEC: $_\n" } @sections;
    @sections;
}

sub _parse_log_rev {
    my ($self, $text) = @_;
    my ($rev_line, $blurb, @reason) = split /\n/, $text;
    my %info = map {
        split /:\s+/
    } split /;\s*/, $blurb;
    my ($junk, $rev) = split /\s+/, $rev_line;
    $info{'revision'} = $rev;
    $info{'reason'} = \@reason;
#print "REASON: @reason\n";
#map { print "$_ => $info{$_}\n" } keys %info;
    \%info;
}

sub _parse_log_header {
    my ($self, $text) = @_;
    $text =~ s#(description:.*)##s;
    my $desc = join "\n ", split /\n/, $1;
    $text .= $desc;
    my @parts = $text =~ /^(\S.*?)(?=^\S|\Z)/gms;
    chomp @parts;
#map { print "PART: $_\n" } @parts;
    my %info = map {
        split /:\s*/, $_, 2
    } @parts;
#map { print "$_ => $info{$_}\n" } keys %info;
    \%info;
}

sub _read_pipe {
    my ($self, $cmd) = @_;
    local *PIPE;
    open PIPE, $cmd;
    local $/ = undef;
    my $contents = <PIPE>;
    close PIPE;
    $contents = '' unless defined $contents;
    return $contents;
}

1;


__END__

=head1 NAME

VCS::Cvs - notes for the CVS implementation

=head1 SYNOPSIS

    $ENV{CVSROOT} = '/cvsroot';
    use VCS;
    $file = VCS::File->new('vcs://localhost/VCS::Cvs/source/project/Makefile');

=head1 DESCRIPTION

Currently, the user needs to ensure that the environmental requirements
for CVS command line tools are satisfied.

=head1 AVAILABILITY

VCS::Cvs is currently part of the main VCS distribution.

=head1 COPYRIGHT

Copyright (c) 1998-2003 Leon Brocard & Greg McCarroll. All rights
reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<VCS>.

=cut


VCS/Rcs.pm
package VCS::Rcs;

use strict;
use vars qw($VERSION $LOG_CMD %LOG_CACHE);
use VCS::Rcs::Dir;
use VCS::Rcs::File;
use VCS::Rcs::Version;

$VERSION = '0.06';

$LOG_CMD = "rlog";

sub _boiler_plate_info {
    my ($self, $what) = @_;
    my ($header, $log) = $self->_split_log($self->{VERSION});
    my $rev_info = $self->_parse_log_rev($log);
    $rev_info->{$what};
}

sub _split_log {
    my ($self, $version) = @_;
    my $log_text;
    my $cache_id = $self->url;
    unless (defined($log_text = $LOG_CACHE{$cache_id})) {
        my $cmd =
            $LOG_CMD .
            (defined $version ? " -r$version" : '') .
            " " . $self->path . " |";
        $LOG_CACHE{$cache_id} = $log_text = $self->_read_pipe($cmd);
    }
    my @sections = split /\n[=\-]+\n/, $log_text;
    @sections = ($sections[0], grep {/^revision $version(\n|\s)/} @sections) if $version;
#map { print "SEC: $_\n" } @sections;
    @sections;
}

sub _parse_log_rev {
    my ($self, $text) = @_;
    my ($rev_line, $blurb, @reason) = split /\n/, $text;
    my %info = map {
        split /:\s+/
    } split /;\s*/, $blurb;
    my ($junk, $rev) = split /\s+/, $rev_line;
    $info{'revision'} = $rev;
    $info{'reason'} = \@reason;
#print "REASON: @reason\n";
#map { print "$_ => $info{$_}\n" } keys %info;
    \%info;
}

sub _parse_log_header {
    my ($self, $text) = @_;
    $text =~ s#(description:.*)##s;
    my $desc = join "\n ", split /\n/, $1;
    $text .= $desc;
    my @parts = $text =~ /^(\S.*?)(?=^\S|\Z)/gms;
    chomp @parts;
#map { print "PART: $_\n" } @parts;
    my %info = map {
        split /:\s*/, $_, 2
    } @parts;
#map { print "$_ => $info{$_}\n" } keys %info;
    \%info;
}

sub _read_pipe {
    my ($self, $cmd) = @_;
#use Carp; Carp::cluck "r_p: $cmd\n";
    local *PIPE;
    open PIPE, $cmd;
    local $/ = undef;
    my $contents = <PIPE>;
    close PIPE;
    return $contents;
}

1;

=head1 NAME

VCS::Rcs - notes for the rcs implementation

=head1 SYNOPSIS

    use VCS;
    $file = VCS::File->new('/source/rcsrepos/project/Makefile');

=head1 DESCRIPTION

Currently, the user needs to ensure that their environment has the
B<rcs> toolset available, including B<rlog>, B<rcsdiff>, B<co>, et al.
On Unix like environments ensure that the C<$PATH> environment variable
has the appropriate directory listed.  On Windows be sure that the C<%PATH%>
variable has the directory with B<rlog.exe> etc. in it.

=head1 AVAILABILITY

VCS::Rcs is currently part of the main VCS distribution.

=head1 COPYRIGHT

Copyright (c) 1998-2003 Leon Brocard & Greg McCarroll. All rights
reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<VCS>.

=cut


Makefile.PL
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my $exe_type = '';
my $EXE_FILES = [qw(examples/diff-hist)];
if ($^O eq 'VMS') {
   $EXE_FILES = [qw([.examples]diff-hist.com)];
   $exe_type = '.com';
}
if ($^O eq 'os2' || $^O eq 'MSWin32') {
   $EXE_FILES = [qw(examples/diff-hist.bat)];
   $exe_type = '.bat';
}
my $PL_FILES = {('examples/diff-hist.PL' => "examples/diff-hist$exe_type")};

WriteMakefile(
    'NAME'	    => 'VCS',
    'ABSTRACT_FROM' => 'VCS.pm', # finds =head1 NAME line for PPD
    'VERSION_FROM'  => 'VCS.pm', # finds $VERSION
    'PREREQ_PM'     => { 'URI' => 0 },
    'PL_FILES'      => $PL_FILES,
    'EXE_FILES'     => $EXE_FILES,
    'dist'          => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
);


examples/diff-hist.PL
#!/usr/local/bin/perl

# standard perl modules to assist with .PL extraction
use Config;
use File::Basename qw(&basename &dirname);
use Cwd;

my $origdir = cwd();
chdir dirname($0);
my $name = basename($0, '.PL');
   $name .= '.com' if $^O eq 'VMS';
   $name .= '.bat' if $^O eq 'MSWin32';
open EXE ,">$name" or die "Cannot create $name: $!";
print "Extracting $name (with variable substitutions)\n";
print EXE << "!GROK!THIS!";
$Config{'startperl'}
!GROK!THIS!
if ($^O eq 'VMS') {
print EXE <<'!NO!SUBS!';
#!/perl
!NO!SUBS!
}
print EXE << '!NO!SUBS!';

$| = 1;

#use FindBin;
#use lib "$FindBin::Bin/../";
use VCS;

if (scalar(@ARGV) == 0) {
    print << "EOUSAGE";
Usage:

    diff-hist file [file...]
EOUSAGE
    exit;
}


foreach my $arg (@ARGV) {
    my $file = VCS::File->new($arg);
    my @versions = $file->versions;
    my @diff_pairs;
    for (my $count = @versions - 1; $count > 0; $count--) {
        push @diff_pairs, [ $versions[$count - 1], $versions[$count] ];
    }
    map {
        my $old = $_->[0];
        my $new = $_->[1];
        print
            '*** Changes from version ',
            $old->version,
            ' to version ',
            $new->version,
            ' (',
            $new->date,
            ")\n",
            'What: ', $new->path, "\n",
            'Version: ', $new->version, "\n",
            'When: ', $new->date, "\n",
            'Who: ', $new->author, "\n",
            'Tags: ', (join "\n      ", $new->tags), "\n",
            'Why: ', (join "\n     ", $new->reason), "\n",
            $old->diff($new)
    } @diff_pairs;
}

__END__

=head1 NAME

diff-hist - view revision history of file under version control

=head1 SYNOPSIS

diff-hist file [file...]

=head1 DESCRIPTION

For each file specified, the differences (in diff -u2 format) are shown
between each successive version back to the first, in reverse
chronological order.

=head1 SEE ALSO

L<VCS>.

=cut
!NO!SUBS!


VCS.pm
package VCS;

my @IMPLEMENTATIONS;
my $CONTAINER_PAT = '(' . join('|', qw(VCS/Dir VCS/File VCS/Version)) . ')';

use vars qw($VERSION);
use VCS::Dir;
use VCS::File;
use VCS::Version;
use URI;

$VERSION = '0.16';

sub parse_url {
    # vcs://hostname/classname/...
    my ($class, $url) = @_;
    my $uri = URI->new($url);
    die "Non-vcs URL '$url' passed!\n" unless $uri->scheme eq 'vcs';
    my $path = $uri->path;
    $path =~ s#^/([^/]+)##;
    my $classname = $1;
    ($uri->authority, $classname, $path, $uri->query)
}

sub _class2file {
    my $class = shift;
    $class =~ s#::#/#g;
    $class .= '.pm';
    $class;
}

sub class_load {
    my ($class, $to_load) = @_;
    require(_class2file($to_load));
}

1;

__END__

=head1 NAME

VCS - (OBSOLETE, USE VCI INSTEAD) Version Control System access in Perl

=head1 SYNOPSIS

    use VCS;
    $file = VCS::File->new($ARGV[0]);
    print $file->url, ":\n";
    for $version ($file->versions) {
        print $version->version,
              ' was checked in by ',
              $version->author,
              "\n";
    }

=head1 DESCRIPTION

B<NOTE:> This module has been unmaintained since 2004. It is recommended that
you use the L<VCI> module instead, which is currently maintained, supports
many more VCSes, and has more features.

C<VCS> is an API for abstracting access to all version control systems
from Perl code. This is achieved in a similar fashion to the C<DBI>
suite of modules. There are "container" classes, C<VCS::Dir>,
C<VCS::File>, and C<VCS::Version>, and "implementation" classes, such
as C<VCS::Cvs::Dir>, C<VCS::Cvs::File>, and C<VCS::Cvs::Version>, which
are subclasses of their respective "container" classes.

The container classes are instantiated with URLs. There is a URL scheme
for entities under version control. The format is as follows:

    vcs://localhost/VCS::Cvs/fs/path/?query=1

The "query" part is ignored for now. The path must be an absolute path,
meaningful to the given class. The class is an implementation class,
such as C<VCS::Cvs>.

The "container" classes work as follows: when the C<new> method of a
container class is called, it will parse the given URL, using the
C<VCS-E<gt>parse_url> method. It will then call the C<new> of the
implementation's appropriate container subclass, and return the
result. For example,

    VCS::Version->new('vcs://localhost/VCS::Cvs/fs/path/file/1.2');

will return a C<VCS::Cvs::Version>.

An implementation class is recognised as follows: its name starts with
C<VCS::>, and C<require "VCS/Classname.pm"> will load the appropriate
implementation classes corresponding to the container classes.

=head1 VCS METHODS

=head2 VCS-E<gt>parse_url

This returns a four-element list:

    ($hostname, $classname, $path, $query)

For example,

    VCS->parse_url('vcs://localhost/VCS::Cvs/fs/path/file/1.2');

will return

    (
        'localhost',
        'VCS::Cvs',
        '/fs/path/file/1.2',
        ''
    )

This is mostly intended for use by the container classes, and its
interface is subject to change.

=head2 VCS-E<gt>class_load

This loads its given implementation class.

This is mostly intended for use by the container classes, and its
interface is subject to change.

=head1 VCS::* METHODS

Please refer to the documentation for L<VCS::Dir>, L<VCS::File>,
and L<VCS::Version>; as well as the implementation specific documentation
as in L<VCS::Cvs>, L<VCS::Rcs>.

=head1 AVAILABILITY

Much of this information is incorrect, the current up to date
version of VCS is held on a different CVS server now, i'm going
to make things a little more public and then update the below
information - thanks for your patience.

VCS.pm and its friends are available from CPAN.  There is a web page
at:

    http://www.astray.com/VCS/

as well as a sourceforge project page at:

    http://sourceforge.net/projects/vcs/

=head1 MAILING LIST

There is currently a mailing list about VCS. Go to the following
webpage to subscribe to it:

    http://www.astray.com/mailman/listinfo/vcs

There is a web archive of the mailing list at:

    http://www.astray.com/pipermail/vcs/

General queries should be made directly to the mailing list.

=head1 AUTHORS

Greg McCarroll <greg@mccarroll.org.uk>
Leon Brocard

=head1 KUDOS

Thanks to the following for patches,

    Richard Clamp
    Pierre Denis
    Slaven Rezic


=head1 COPYRIGHT

Copyright (c) 1998-2003 Leon Brocard & Greg McCarroll. All rights
reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<VCI>

L<VCS::Cvs>, L<VCS::Dir>, L<VCS::File>, L<VCS::Rcs>, L<VCS::Version>.
